### PR TITLE
Rework slice recursive freeze/thaw

### PR DESCRIPTION
**Pull request by AdrianVovk**
Fixes https://github.com/systemd/systemd/issues/30640
Fixes https://github.com/systemd/systemd/issues/15850
Fixes https://github.com/systemd/systemd/issues/15849

### CODE REVIEW

**Comment by AdrianVovk** [comment]
This also introduces freeze/thaw for mount units via FIFREEZE and FITHAW, but it's really just a proof of concept to show what becomes possible w/ the rework. Unsure if this is something you'd want :shrug:

**Comment by AdrianVovk** [comment]
OK, I think this is ready for a look

CC @msekletar, @Werkov, and @poettering

Also, CC @msizanoen1. I've effectively reverted a14137d90e5f50ad8627c85ae94731a5c9948227. Please take a look

File: src/basic/unit-def.c [hunk 1]
```diff
@@ -117,14 +117,30 @@ static const char* const unit_active_state_table[_UNIT_ACTIVE_STATE_MAX] = {
 DEFINE_STRING_TABLE_LOOKUP(unit_active_state, UnitActiveState);
 
 static const char* const freezer_state_table[_FREEZER_STATE_MAX] = {
-        [FREEZER_RUNNING]  = "running",
-        [FREEZER_FREEZING] = "freezing",
-        [FREEZER_FROZEN]   = "frozen",
-        [FREEZER_THAWING]  = "thawing",
+        [FREEZER_RUNNING]         = "running",
+        [FREEZER_FREEZING]        = "freezing",
+        [FREEZER_PARENT_FREEZING] = "parent-freezing",
```
Comment (poettering) [inline src/basic/unit-def.c]: maybe name this FREEZER_FREEZING_BY_PARENT

File: src/basic/unit-def.c [hunk 2]
```diff
@@ -117,14 +117,30 @@ static const char* const unit_active_state_table[_UNIT_ACTIVE_STATE_MAX] = {
...
+        [FREEZER_PARENT_FROZEN]   = "parent-frozen",
+        [FREEZER_THAWING]         = "thawing",
 };
 
 DEFINE_STRING_TABLE_LOOKUP(freezer_state, FreezerState);
 
+static const FreezerState freezer_state_finish_table[_FREEZER_STATE_MAX] = {
+        [FREEZER_RUNNING]         = FREEZER_RUNNING,
+        [FREEZER_FREEZING]        = FREEZER_FROZEN,
+        [FREEZER_PARENT_FREEZING] = FREEZER_PARENT_FROZEN,
```
Comment (poettering) [inline src/basic/unit-def.c]: can you add a comment to the end of these lines explaing briefly what this is about: i.e. `/* freezing due to user request */` and `/* freezing as result of parent slice freezing */`

File: src/basic/unit-def.c [hunk 3]
```diff
@@ -117,14 +117,30 @@ static const char* const unit_active_state_table[_UNIT_ACTIVE_STATE_MAX] = {
...
+        [FREEZER_FROZEN]          = FREEZER_FROZEN,
+        [FREEZER_PARENT_FROZEN]   = FREEZER_PARENT_FROZEN,
+        [FREEZER_THAWING]         = FREEZER_RUNNING,
+};
+
+FreezerState freezer_state_finish(FreezerState state) {
+        assert(state >= 0);
```
Comment (poettering) [inline src/basic/unit-def.c]: also `assert(state < _FREEZER_STATE_MAX);`

File: src/core/cgroup.c [hunk 4]
```diff
@@ -3897,8 +3897,10 @@ static int unit_check_cgroup_events(Unit *u) {
                         unit_add_to_cgroup_empty_queue(u);
         }
 
-        /* Disregard freezer state changes due to operations not initiated by us */
-        if (values[1] && IN_SET(u->freezer_state, FREEZER_FREEZING, FREEZER_THAWING)) {
+        /* Disregard freezer state changes due to operations not initiated by us.
+         * See: https://github.com/systemd/systemd/pull/13512/files#r416469963 and
+         *      https://github.com/systemd/systemd/pull/13512#issuecomment-573007207 */
+        if (values[1] && IN_SET(u->freezer_state, FREEZER_FREEZING, FREEZER_PARENT_FREEZING, FREEZER_THAWING)) {
```
Comment (poettering) [inline src/core/cgroup.c]: what about PARENT_THAWING?

File: src/core/cgroup.c [hunk 5]
```diff
@@ -4891,66 +4893,82 @@ void manager_invalidate_startup_units(Manager *m) {
...
-        /* Ignore all requests to thaw init.scope or -.slice and reject all requests to freeze them */
-        if (unit_has_name(u, SPECIAL_ROOT_SLICE) || unit_has_name(u, SPECIAL_INIT_SCOPE))
-                return action == FREEZER_FREEZE ? -EPERM : 0;
+        if (streq(val, "0"))
+                *ret = FREEZER_RUNNING;
+        else if (streq(val, "1"))
+                *ret = FREEZER_FROZEN;
+        else
+                log_unit_debug_errno(u, SYNTHETIC_ERRNO(EINVAL), "Unexpected cgroup frozen state: %s", val);
+        return 0;
```
Comment (poettering) [inline src/core/cgroup.c:4921]: shouldn't return 0 (i.e. success) without initializing the return paramer, as per our coding style that says all ret params must be initialized on success, and left untouched on failure

File: src/core/dbus-unit.c [hunk 6]
```diff
@@ -723,21 +723,14 @@ int bus_unit_method_clean(sd_bus_message *message, void *userdata, sd_bus_error
...
         assert(IN_SET(action, FREEZER_FREEZE, FREEZER_THAW));
 
-        if (action == FREEZER_FREEZE) {
-                perm = "stop";
-                method = unit_freeze;
-        } else {
-                perm = "start";
-                method = unit_thaw;
-        }
+        perm = (action == FREEZER_FREEZE) ? "stop" : "start";
```
Comment (poettering) [inline src/core/dbus-unit.c]: the () are unnecessary

File: src/core/job.h [hunk 7]
```diff
@@ -96,6 +96,7 @@ enum JobResult {
         JOB_UNSUPPORTED,         /* Couldn't start a unit, because the unit type is not supported on the system */
         JOB_COLLECTED,           /* Job was garbage collected, since nothing needed it anymore */
         JOB_ONCE,                /* Unit was started before, and hence can't be started again */
+        JOB_FROZEN,              /* Unit is currently frozen, so we can't safely operate on it */
```
Comment (poettering) [inline src/core/job.h:99]: oh, nice, this is a good idea

File: src/core/mount.c [hunk 8]
```diff
@@ -2458,6 +2461,58 @@ static const char* const mount_result_table[_MOUNT_RESULT_MAX] = {
...
+                              FREEZER_THAW, FREEZER_PARENT_THAW));
+
+        unit_next_freezer_state(u, action, &next, &target);
+        next = freezer_state_finish(next); /* We're completely sync */
+
+        log_unit_debug(u, "Mount freezer state was %s, now %s.",
+                       freezer_state_to_string(u->freezer_state),
+                       freezer_state_to_string(next));
+
+        fd = open(m->where, O_RDONLY|O_NOCTTY);
```
Comment (poettering) [inline src/core/mount.c]: O_CLOEXEC

Comment (poettering) [reply src/core/mount.c]: O_DIRECTORY

File: src/core/mount.c [hunk 9]
```diff
@@ -2458,6 +2461,58 @@ static const char* const mount_result_table[_MOUNT_RESULT_MAX] = {
...
+        FreezerState next, target;
+        int r;
+
+        assert(u);
+        assert(m);
+        assert(IN_SET(action, FREEZER_FREEZE, FREEZER_PARENT_FREEZE,
+                              FREEZER_THAW, FREEZER_PARENT_THAW));
+
+        unit_next_freezer_state(u, action, &next, &target);
+        next = freezer_state_finish(next); /* We're completely sync */
```
Comment (poettering) [inline src/core/mount.c]: the ioctl is indeed sync, like all disk accesses. we could theoroetically do this async, i.e. fork off a thread or so. but maybe it's not worth the effort for now, given that disk accesses are pretty much always sync, dunno.

File: src/core/mount.c [hunk 10]
```diff
@@ -2458,6 +2461,58 @@ static const char* const mount_result_table[_MOUNT_RESULT_MAX] = {
...
+                return -errno;
+
+        if (target == FREEZER_FROZEN) {
+                r = RET_NERRNO(ioctl(fd, FIFREEZE, 0));
+                if (r == -EBUSY) /* Already frozen by us */
+                        return 0;
+        } else {
+                r = RET_NERRNO(ioctl(fd, FITHAW, 0));
+                if (r == -EINVAL) /* Not frozen by us */
+                        return 0;
```
Comment (poettering) [inline src/core/mount.c]: logging is sometimes a bit wonky in pid 1, and needs some clean-ups, but i think we should do log_unit_warning() here for unexpected freeze/thaw errors at least.

**Comment by poettering** [comment]
so, the last commit, what's your thinking there precisely, how to you intend to make use of this?

File: src/core/cgroup.c (hunk 4 above)
Comment (AdrianVovk) [reply src/core/cgroup.c]: There's no such state

**Comment by AdrianVovk** [comment]
> last commit, what's your thinking there precisely

Part of the motivation for the rework is to clean up the code handling freezing/thawing, which has the effect of making the behavior sane enough that other kinds of units can be frozen/thawed (i.e. not via cgroup) should the use-case arise. The toy example I had in my head for this was a mount unit frozen/thawed via FIFREEZE/FITHAW. It was simple to throw together and so I did as a proof of concept.

> how to you intend to make use of this

I don't have a use-case for it. I'm perfectly happy to drop the commit

File: src/core/job.c [hunk 11]
```diff
@@ -633,16 +633,19 @@ static const char* job_done_message_format(Unit *u, JobType t, JobResult result)
                 [JOB_UNSUPPORTED] = "Starting of %s unsupported.",
                 [JOB_COLLECTED]   = "Unnecessary job was removed for %s.",
                 [JOB_ONCE]        = "Unit %s has been started before and cannot be started again.",
+                [JOB_FROZEN]      = "Can not start frozen unit %s.",
         };
         static const char* const generic_finished_stop_job[_JOB_RESULT_MAX] = {
                 [JOB_DONE]        = "Stopped %s.",
                 [JOB_FAILED]      = "Stopped %s with error.",
                 [JOB_TIMEOUT]     = "Timed out stopping %s.",
+                [JOB_FROZEN]      = "Can not stop frozen unit %s.",
```
Comment (YHNdnzj) [inline src/core/job.c]: "Cannot"

File: src/core/job.c [hunk 12]
```diff
@@ -633,16 +633,19 @@ static const char* job_done_message_format(Unit *u, JobType t, JobResult result)
                 [JOB_UNSUPPORTED] = "Starting of %s unsupported.",
                 [JOB_COLLECTED]   = "Unnecessary job was removed for %s.",
                 [JOB_ONCE]        = "Unit %s has been started before and cannot be started again.",
+                [JOB_FROZEN]      = "Can not start frozen unit %s.",
```
Comment (YHNdnzj) [inline src/core/job.c]: Also here

File: src/core/job.c [hunk 13]
```diff
@@ -633,16 +633,19 @@ static const char* job_done_message_format(Unit *u, JobType t, JobResult result)
...
                 [JOB_DONE]        = "Stopped %s.",
                 [JOB_FAILED]      = "Stopped %s with error.",
                 [JOB_TIMEOUT]     = "Timed out stopping %s.",
+                [JOB_FROZEN]      = "Can not stop frozen unit %s.",
         };
         static const char* const generic_finished_reload_job[_JOB_RESULT_MAX] = {
                 [JOB_DONE]        = "Reloaded %s.",
                 [JOB_FAILED]      = "Reload failed for %s.",
                 [JOB_TIMEOUT]     = "Timed out reloading %s.",
+                [JOB_FROZEN]      = "Can not reload frozen unit %s.",
```
Comment (YHNdnzj) [inline src/core/job.c]: And here.

File: src/core/job.c [hunk 14]
```diff
@@ -704,6 +707,7 @@ static const struct {
         [JOB_UNSUPPORTED] = { LOG_WARNING, ANSI_HIGHLIGHT_YELLOW, "UNSUPP" },
         [JOB_COLLECTED]   = { LOG_INFO,                                    },
         [JOB_ONCE]        = { LOG_ERR,     ANSI_HIGHLIGHT_RED,    " ONCE " },
+        [JOB_FROZEN]      = { LOG_ERR,     ANSI_HIGHLIGHT_RED,    "FROZEN" },
```
Comment (YHNdnzj) [inline src/core/job.c:710]: Please also update `check_wait_response` in `shared/bus-wait-for-jobs.c` to include the new job result.

**Review by YHNdnzj — CHANGES_REQUESTED** [review]

File: src/basic/mountpoint-util.c [hunk 15]
```diff
@@ -535,6 +535,25 @@ bool fstype_can_uid_gid(const char *fstype) {
                           "vfat");
 }
 
+bool fstype_can_freeze_thaw(const char *fstype) {
```
Comment (YHNdnzj) [inline src/basic/mountpoint-util.c]: assert(fstype)

File: src/core/mount.c [hunk 16]
```diff
@@ -2458,6 +2461,58 @@ static const char* const mount_result_table[_MOUNT_RESULT_MAX] = {
...
+        fstype = mount_get_fstype(m);
+        if (!fstype)
+                return false;
+
+        return fstype_can_freeze_thaw(fstype);
+}
+
+static int mount_freezer_action(Unit *u, FreezerAction action) {
+        _cleanup_close_ int fd = -EBADF;
+        Mount *m = MOUNT(u);
```
Comment (YHNdnzj) [inline src/core/mount.c]: Please just use ASSERT_PTR

File: src/core/mount.c [hunk 17]
```diff
@@ -2458,6 +2461,58 @@ static const char* const mount_result_table[_MOUNT_RESULT_MAX] = {
...
+        FreezerState next, target;
+        int r;
+
+        assert(u);
+        assert(m);
```
Comment (YHNdnzj) [inline src/core/mount.c]: And drop these

File: src/core/cgroup.c (hunk 4 above)
Comment (poettering) [reply src/core/cgroup.c]: ah right

**Comment by poettering** [comment]
> I don't have a use-case for it. I'm perfectly happy to drop the commit

so, i am torn. I kinda like it, but note that mount units can also be assigned a slice (and by default are assigned system.slice). We do that because of userspace mount tools, i.e. fuse mounts. That's why mounts also have a cgroup and everything: the processes backing those mounts might run continously. Now if a slice is frozen that has mounts in it, is it right that this also freezes the file systems? 

i have no clear answer to this, I am a bit unsure.

I mean, on one hand: if a mount is indeed a fuse mount or something else that forks off a bg process, then freezing this process would typically mean that the fs is effectively frozen too, much like what your patch does for block based file systems...

so i think, if we want to be fully correct we'd freeze the cgroup *and* the block device for a .mount unit... but this is also problematic, since we ourselves might end up access the path so we might deadlock ourselves?

I mean, how does that work anyway? if the open() we do to issue the thaw ioctl accesses the fs and the fs is frozen, how can we be sure it won't deadlock? should we keep the fd open when we freeze so that we do not have to reopen, but can just fire one more ioctl just like that?

File: src/core/mount.c (hunk 8 above)
Comment (AdrianVovk) [reply src/core/mount.c]: > O_DIRECTORY

We may be bind-mounting over a file, right? So we can't do O_DIRECTORY

Comment (AdrianVovk) [reply src/core/mount.c]: Though I suppose you can't actually freeze bind mounts, so you're right. O_DIRECTORY should be there

File: src/core/mount.c (hunk 9 above)
Comment (AdrianVovk) [reply src/core/mount.c]: As far as I can tell, filesystem freezing should be instant; I don't think it recurses down the tree like a cgroup freeze does. But of course the implementation is filesystem dependent, so :shrug: 

I think it makes most sense to leave it sync for now, and then if someone discovers a filesystem that does actually recursively do things in response to a freeze request then we make it async

CC @brauner

File: src/core/mount.c (hunk 10 above)
Comment (AdrianVovk) [reply src/core/mount.c]: I did log_unit_error instead, since we don't ignore the exit value and do report failure to our caller

**Comment by AdrianVovk** [comment]
`FIFREEZE` doesn't prevent reading from the filesystem; it only prevents writing it. It's not like a dm-crypt suspend where reads are also suspended forever. So the path should always be accessible enough to `open()` it

Fuse does complicate matters a little, because the process that's serving the `open()` request will itself be frozen. But I think that's manageable: we should just thaw the cgroup before we try to thaw the FS via ioctl. When we freeze we'd do the opposite: freeze via ioctl before freezing the cgroup.

But also: there are no FUSE filesystems on the list we support freezing with, so this may be a non-issue. At least until some FUSE filesystem is added to the list

**Comment by brauner** [comment]
> I mean, how does that work anyway? if the open() we do to issue the thaw ioctl accesses the fs and the fs is frozen, how can we be sure it won't deadlock? should we keep the fd open when we freeze so that we do not have to reopen, but can just fire one more ioctl just like that?

Of course, the VFS layer is designed in a manner that prevents you from deadlocking just by issuing an FIFREEZE or FITHAW ioctl using an file descriptor.

**Comment by poettering** [comment]
> Of course, the VFS layer is designed in a manner that prevents you from deadlocking just by issuing an FIFREEZE or FITHAW ioctl using an file descriptor.

Right, but that requires us to keep a fd open, right? it would not be safe to start again with open() and a path to a frozen fs, right?

**Comment by poettering** [comment]
> `FIFREEZE` doesn't prevent reading from the filesystem; it only prevents writing it. It's not like a dm-crypt suspend where reads are also suspended forever. So the path should always be accessible enough to `open()` it

Hmm, I thought that this is is now propagate downwards and `FIFREEZE` actually does also result in a block layer suspend if the mechanism exists there. @brauner did i misunderstand that?

**Comment by poettering** [comment]
Looks good to merge if you drop the last commit.

**Comment by AdrianVovk** [comment]
OK I dropped d09f2c51261be43a9950cc7ce63876fd394e5de2

I had also broken the tests by renaming the state, so I fixed that also (at least I think. We'll have to see what CI says)

**Comment by AdrianVovk** [comment]
CI failures appear unrelated

The various `jammy-*` tests all fail in `TEST-75-RESOLVED`

`testing-farm:fedora-rawhide-x86_64 ` is failing due to packaging conflicts

**Comment by AdrianVovk** [comment]
Thanks!

**Pull request merged**